### PR TITLE
Support for GitHub Enterprise closes #70

### LINKED ISSUE #70: Support for GitHub Enterprise
It would be great if the GitHub plugin supported GitHub Enterprise. This would be mostly a change to the config and connection setup to allow a custom host etc.

After development, we'll require someone with access to GitHub Enterprise to help with final testing.

### PR DESCRIPTION
**Pull request by mheiges**
# Example query results
<details>
  <summary>Results</summary>
  
```
Add example SQL query results here (please include the input queries as well)
```
</details>

### CODE REVIEW

**Comment by e-gineer** [comment]
Quick note that I'm excited to see support for GitHub enterprise - thanks @mheiges! Have wanted this for a long time, but didn't have the ability to test.

**Comment by mheiges** [comment]
some sample queries run against GHE, mostly verifying that `repository_full_name` is correct.


```
steampipe query "select name, repository_full_name from github_search_code  where query = 'filename:build.Dockerfile path:automation/docker'"

+------------------+-------------------------------------------+
| name             | repository_full_name                      |
+------------------+-------------------------------------------+
| build.Dockerfile | myORG/myorg-ui                            |
| build.Dockerfile | myORG/myorg-rpm-service                   |
| build.Dockerfile | myORG/myorg-certification-service         |
| build.Dockerfile | myORG/myorg-report-service                |
```

```
steampipe query "select title, repository_full_name from github_search_issue  where query = 'is:open repo:myORG/myorg-foo-service'"
+-------------------------+-----------------------------+
| title                   | repository_full_name        |
+-------------------------+-----------------------------+
| yeah, this is a problem | myORG/myorg-foo-service     |
+-------------------------+-----------------------------+
```


```
steampipe query "select title, repository_full_name from github_search_pull_request  where query = 'is:open repo:myORG/myorg-foo-service'"
+----------------------------+-----------------------------+
| title                      | repository_full_name        |
+----------------------------+-----------------------------+
| build with lockfile        | myORG/myorg-foo-service     |
+----------------------------+-----------------------------+
```

```
steampipe query "select description, repository_full_name from github_search_label where repository_id = 3753 and query = 'bug'"
+-------------------------+-----------------------------+
| description             | repository_full_name        |
+-------------------------+-----------------------------+
| Something isn't working | myORG/myorg-foo-service     |
+-------------------------+-----------------------------+
```

File: github/utils.go [hunk 1]
```diff
@@ -16,24 +18,63 @@ import (
...
+		uv3, err := url.Parse(baseURL)
+		if err != nil {
+			panic(fmt.Sprintf("github.base_url is invalid: %s", baseURL))
+		}
+
+		if uv3.String() != "https://api.github.com/" {
+			uv3.Path = uv3.Path + "api/v3/"
+		}
+
+		conn, err = github.NewEnterpriseClient(uv3.String(), "", tc)
```
Comment (cbruno10) [inline github/utils.go:66]: Is there any future compatibility you're aware of that wouldn't work if we don't pass the upload URL when creating the client?

File: github/utils.go [hunk 2]
```diff
@@ -16,24 +18,63 @@ import (
...
+	// enterprise installs.
+	if baseURL != "" {
+
+		uv3, err := url.Parse(baseURL)
+		if err != nil {
+			panic(fmt.Sprintf("github.base_url is invalid: %s", baseURL))
+		}
+
+		if uv3.String() != "https://api.github.com/" {
+			uv3.Path = uv3.Path + "api/v3/"
```
Comment (cbruno10) [inline github/utils.go:63]: Do we need to include this, or does the go-github library automatically attach this suffix, per https://github.com/google/go-github/blob/a58d5f0adf74ef73aadaac476f7285afc77a3b43/github/github.go#L357-L364

File: github/utils.go (hunk 1 above)
Comment (cbruno10) [inline github/utils.go:66]: If someone sets `base_url` as https://github.com/api/v3`, does the enterprise client still work, or does it fail/have another behaviour?

File: github/table_github_search_code.go [hunk 3]
```diff
@@ -117,9 +116,8 @@ func tableGitHubSearchCodeList(ctx context.Context, d *plugin.QueryData, h *plug
 func extractSearchCodeRepositoryFullName(_ context.Context, d *transform.TransformData) (interface{}, error) {
 	code := d.HydrateItem.(*github.CodeResult)
 	if code.HTMLURL != nil {
-		rx := regexp.MustCompile(`(?s)` + regexp.QuoteMeta("github.com/") + `(.*?)` + regexp.QuoteMeta("/blob"))
-		replacer := strings.NewReplacer("github.com/", "", "/blob", "")
-		return replacer.Replace(rx.FindString(*code.HTMLURL)), nil
+		rx := regexp.MustCompile(`https?://.+?/(.+?)/blob`)
```
Comment (cbruno10) [inline github/table_github_search_code.go:119]: Have you tested this change (along with other `github_search_*` table changes) against a GitHub cloud instance for backward compatibility? I saw some query results in https://github.com/turbot/steampipe-plugin-github/pull/160#issuecomment-1119199458, but wasn't sure if those were for an enterprise or cloud instance.

**Review by cbruno10 — CHANGES_REQUESTED** [review]
@mheiges Sorry for the slight delay in reviewing this PR, I've added a few questions/suggestions, can you please have a look when you get a chance? Thanks!

File: github/table_github_search_code.go (hunk 3 above)
Comment (mheiges) [reply github/table_github_search_code.go:119]: the tests were against github.com (`base_url = "https://api.github.com/`) and GH Enterprise with custom hostname (`base_url = "https://abc.xyz.gov/`. I guess our GHE is Server, not Cloud - I'm not clear on the difference.

File: github/utils.go (hunk 1 above)
Comment (mheiges) [reply github/utils.go:66]: `base_url = "https://api.github.com/` works.

`base_url = https://github.com/api/v3` returns a `404` status for `https://github.com/api/v3/api/v3/search/issues`

Comment (mheiges) [reply github/utils.go:63]: I don't know. This came from the github-enterprise-support branch, I haven't explored the client requirements.

**Review by cbruno10 — APPROVED** [review]

**Comment by cbruno10** [comment]
@mheiges Thank you for testing out this branch and making improvements to it as well! We've just released v0.15.0 of this plugin, which includes this PR. If you see any issues, please let us know.

**Pull request merged**